### ISSUE TITLE
tf.nn.softmax_cross_entropy_with_logits_v2 returns wrong value with soft labels

### ISSUE THREAD

**Issue by fenghou1st** [tensorflow/tensorflow#21271]
### System information
- **Have I written custom code**: Yes
- **OS Platform and Distribution**: Linux CentOS 7.4
- **Mobile device if the issue happens on mobile device**: N/A
- **TensorFlow installed from**: binary
- **TensorFlow version**: v1.9.0-0-g25c197e023 1.9.0
- **Python version**: Python 3.6.3
- **Bazel version**: N/A
- **GCC/Compiler version**: N/A
- **CUDA/cuDNN version**: 9.0 / 7.0.5
- **GPU model and memory**:  GeForce 940MX, 4GB
- **Exact command to reproduce**: save the codes below as a .py file, and run it with command-line something like `python3 test.py`.

### Describe the problem
I use soft labels (for example, [0.2, 0.8] instead of [0, 1]) in a CNN model, in which I use `tf.nn.softmax_cross_entropy_with_logits_v2` for loss computing. But when I trained the model, the loss became +inf in 10 steps, so I debugged the codes and found that the problem was caused by `tf.nn.softmax_cross_entropy_with_logits_v2`.
So I implemented the softmax and cross_entropy process separately, then the returned value seemed to make sense. 

### Source code / logs
```python
import numpy as np
import tensorflow as tf

tf.enable_eager_execution()
tf.executing_eagerly()

logits = [[-4885.4614, 4878.027], [-5188.321, 5179.7915], [-4121.558, 4114.9995], [-5165.612, 5157.5044], [-4152.7183, 4145.978], [-5175.428, 5167.603], [-4514.224, 4506.477], [-4752.5854, 4745.2524], [-5580.9463, 5572.2275], [-5164.766, 5156.6685], [-4273.686, 4266.31], [-4886.724, 4879.5757], [-5216.2935, 5208.269], [-5411.082, 5402.344], [-6057.239, 6048.3647], [-5314.882, 5306.708], [-5674.2505, 5664.6436], [-5650.7827, 5642.1997], [-4301.4194, 4294.5957], [-5156.4683, 5148.283], [-5032.6797, 5024.821], [-5072.1533, 5064.013], [-4129.488, 4123.4355], [-4915.1147, 4907.8643], [-5256.5747, 5248.351], [-5297.694, 5289.386], [-4979.939, 4971.691], [-4895.983, 4887.4897], [-4757.732, 4749.2886], [-4871.5654, 4863.604], [-4772.0356, 4763.9414], [-4528.853, 4521.4424],]
labels = [[9.6226019e-01, 3.7739828e-02], [9.5367432e-07, 9.9999905e-01], [9.9017835e-01, 9.8216468e-03], [8.3446503e-07, 9.9999917e-01], [9.9999952e-01, 4.9012118e-07], [3.6466300e-02, 9.6353370e-01], [3.2732385e-01, 6.7267615e-01], [2.1918458e-01, 7.8081542e-01], [3.4707606e-02, 9.6529239e-01], [1.3036132e-03, 9.9869639e-01], [4.1835755e-01, 5.8164245e-01], [5.0599152e-01, 4.9400848e-01], [9.9629784e-01, 3.7021455e-03], [1.9747615e-03, 9.9802524e-01], [8.2850456e-06, 9.9999171e-01], [8.1463850e-01, 1.8536153e-01], [7.6112747e-03, 9.9238873e-01], [9.4729370e-01, 5.2706327e-02], [9.4496381e-01, 5.5036198e-02], [3.3008814e-02, 9.6699119e-01], [1.0292470e-02, 9.8970753e-01], [9.9998099e-01, 1.9039073e-05], [7.5116873e-01, 2.4883127e-01], [9.9973243e-01, 2.6756502e-04], [3.1858683e-04, 9.9968141e-01], [3.6358833e-05, 9.9996364e-01], [9.3631679e-01, 6.3683234e-02], [7.9292524e-01, 2.0707479e-01], [9.9999642e-01, 3.6055078e-06], [4.9336654e-01, 5.0663346e-01], [6.9030523e-03, 9.9309695e-01], [9.9974275e-01, 2.5725813e-04],]

logits = np.array(logits, dtype=np.float32)
labels = np.array(labels, dtype=np.float32)

# Check each row of labels is a valid probability distribution.
labels_sum = tf.reduce_sum(labels, axis=-1)
print(labels_sum)

# Cross entropy computed by tf.nn.softmax_cross_entropy_with_logits_v2
cross_entropy = tf.nn.softmax_cross_entropy_with_logits_v2(labels=labels, logits=logits)
cross_entropy_mean = tf.reduce_mean(cross_entropy)
print(cross_entropy_mean)

# Cross entropy computed separately
softmax = tf.nn.softmax(logits)
cross_entropy_mean = tf.reduce_mean(-tf.reduce_sum(softmax * tf.log(labels), axis=-1))
print(cross_entropy_mean)
```

**Output:**
```
tf.Tensor([1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], shape=(32,), dtype=float32)
tf.Tensor(4538.922, shape=(), dtype=float32)
tf.Tensor(2.621081, shape=(), dtype=float32)
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
Exact command to reproduce
Mobile device

**Comment by fenghou1st** [comment]
My apologies, the cross entropy formula I used above is wrong.
After I correct the error, it returns inf, I think this problem is already solved by `tf.nn.softmax_cross_entropy_with_logits_v2`.
So I read some article which describe the implementation ([Streaming Log-sum-exp Computation](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html)), and implement it.
Then the result is the same as `tf.nn.softmax_cross_entropy_with_logits_v2`.
So the problem is from me, not the TensorFlow code.

**Comment by fenghou1st** [comment]
```python
import numpy as np
import tensorflow as tf

tf.enable_eager_execution()
tf.executing_eagerly()

logits = [[-4885.4614, 4878.027], [-5188.321, 5179.7915], [-4121.558, 4114.9995], [-5165.612, 5157.5044], [-4152.7183, 4145.978], [-5175.428, 5167.603], [-4514.224, 4506.477], [-4752.5854, 4745.2524], [-5580.9463, 5572.2275], [-5164.766, 5156.6685], [-4273.686, 4266.31], [-4886.724, 4879.5757], [-5216.2935, 5208.269], [-5411.082, 5402.344], [-6057.239, 6048.3647], [-5314.882, 5306.708], [-5674.2505, 5664.6436], [-5650.7827, 5642.1997], [-4301.4194, 4294.5957], [-5156.4683, 5148.283], [-5032.6797, 5024.821], [-5072.1533, 5064.013], [-4129.488, 4123.4355], [-4915.1147, 4907.8643], [-5256.5747, 5248.351], [-5297.694, 5289.386], [-4979.939, 4971.691], [-4895.983, 4887.4897], [-4757.732, 4749.2886], [-4871.5654, 4863.604], [-4772.0356, 4763.9414], [-4528.853, 4521.4424],]
labels = [[9.6226019e-01, 3.7739828e-02], [9.5367432e-07, 9.9999905e-01], [9.9017835e-01, 9.8216468e-03], [8.3446503e-07, 9.9999917e-01], [9.9999952e-01, 4.9012118e-07], [3.6466300e-02, 9.6353370e-01], [3.2732385e-01, 6.7267615e-01], [2.1918458e-01, 7.8081542e-01], [3.4707606e-02, 9.6529239e-01], [1.3036132e-03, 9.9869639e-01], [4.1835755e-01, 5.8164245e-01], [5.0599152e-01, 4.9400848e-01], [9.9629784e-01, 3.7021455e-03], [1.9747615e-03, 9.9802524e-01], [8.2850456e-06, 9.9999171e-01], [8.1463850e-01, 1.8536153e-01], [7.6112747e-03, 9.9238873e-01], [9.4729370e-01, 5.2706327e-02], [9.4496381e-01, 5.5036198e-02], [3.3008814e-02, 9.6699119e-01], [1.0292470e-02, 9.8970753e-01], [9.9998099e-01, 1.9039073e-05], [7.5116873e-01, 2.4883127e-01], [9.9973243e-01, 2.6756502e-04], [3.1858683e-04, 9.9968141e-01], [3.6358833e-05, 9.9996364e-01], [9.3631679e-01, 6.3683234e-02], [7.9292524e-01, 2.0707479e-01], [9.9999642e-01, 3.6055078e-06], [4.9336654e-01, 5.0663346e-01], [6.9030523e-03, 9.9309695e-01], [9.9974275e-01, 2.5725813e-04],]

logits = np.array(logits, dtype=np.float32)
labels = np.array(labels, dtype=np.float32)

# Check each row of labels is a valid probability distribution.
labels_sum = tf.reduce_sum(labels, axis=-1)
print(labels_sum)

# Cross entropy computed by tf.nn.softmax_cross_entropy_with_logits_v2
cross_entropy = tf.nn.softmax_cross_entropy_with_logits_v2(labels=labels, logits=logits)
cross_entropy_mean = tf.reduce_mean(cross_entropy)
print(cross_entropy_mean)

# Cross entropy computed separately
softmax = tf.nn.softmax(logits)
cross_entropy_mean = tf.reduce_mean(-tf.reduce_sum(labels * tf.log(softmax), axis=-1))
print(cross_entropy_mean)

# Cross entropy computed with batched log-sum-exp method
alpha = tf.reduce_max(logits, axis=-1, keepdims=True)
log_sum_exp = tf.log(tf.reduce_sum(tf.exp(logits - alpha), axis=-1, keepdims=True)) + alpha
cross_entropy = -tf.reduce_sum((logits - log_sum_exp) * labels, axis=-1)
cross_entropy_mean = tf.reduce_mean(cross_entropy)
print(cross_entropy_mean)
```

**Output:**
```
tf.Tensor([1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.], shape=(32,), dtype=float32)
tf.Tensor(4538.922, shape=(), dtype=float32)
tf.Tensor(inf, shape=(), dtype=float32)
tf.Tensor(4538.922, shape=(), dtype=float32)
```
